copier/copier.go

```diff
@@ -0,0 +1,1367 @@
+				// we were asked to check on this, and it
+				// wasn't a symlink, in which case we'd have
+				// already checked what the link points to
+				result.IsArchive = isArchivePath(globbed)
+			}
+		}
+		// no unskipped matches -> error
+		if len(s.Globbed) == 0 {
+			s.Globbed = nil
+			s.Results = nil
+			s.Error = fmt.Sprintf("copier: stat: %q: %v", glob, syscall.ENOENT)
+		}
+	}
+	return &Response{Stat: StatResponse{Globs: stats}}
+}
+
+func copierHandlerGet(bulkWriter io.Writer, request Request, pm *fileutils.PatternMatcher, idMappings *idtools.IDMappings) (*Response, func() error, error) {
+	statRequest := request
+	statRequest.Request = RequestStat
+	statResponse := copierHandlerStat(request, pm)
+	errorResponse := func(fmtspec string, args ...interface{}) (*Response, func() error, error) {
+		return &Response{Stat: statResponse.Stat, Get: GetResponse{Error: fmt.Sprintf(fmtspec, args...)}}, nil, nil
+	}
+	if statResponse.Stat.Error != "" {
+		return errorResponse("%s", statResponse.Stat.Error)
+	}
+	if len(request.Globs) == 0 {
+		return errorResponse("copier: get: expected at least one glob pattern, got 0")
+	}
+	// build a queue of items by globbing
+	var queue []string
+	globMatchedCount := 0
+	for _, glob := range request.Globs {
+		globMatched, err := filepath.Glob(glob)
+		if err != nil {
+			return errorResponse("copier: get: glob %q: %v", glob, err)
+		}
+		globMatchedCount += len(globMatched)
+		filtered := make([]string, 0, len(globMatched))
+		for _, globbed := range globMatched {
+			rel, excluded, err := pathIsExcluded(request.Root, globbed, pm)
+			if err != nil {
+				return errorResponse("copier: get: checking if %q is excluded: %v", globbed, err)
+			}
+			if rel == "." || !excluded {
+				filtered = append(filtered, globbed)
+			}
+		}
+		if len(filtered) == 0 {
+			return errorResponse("copier: get: glob %q matched nothing (%d filtered out of %v): %v", glob, len(globMatched), globMatched, syscall.ENOENT)
+		}
+		queue = append(queue, filtered...)
+	}
+	// no matches -> error
+	if len(queue) == 0 {
+		return errorResponse("copier: get: globs %v matched nothing (%d filtered out): %v", request.Globs, globMatchedCount, syscall.ENOENT)
+	}
+	cb := func() error {
+		tw := tar.NewWriter(bulkWriter)
+		defer tw.Close()
+		hardlinkChecker := new(util.HardlinkChecker)
+		itemsCopied := 0
+		for i, item := range queue {
+			// if we're not discarding the names of individual directories, keep track of this one
+			relNamePrefix := ""
+			if request.GetOptions.KeepDirectoryNames {
+				relNamePrefix = filepath.Base(item)
+			}
+			// if the named thing-to-read is a symlink, dereference it
+			info, err := os.Lstat(item)
+			if err != nil {
+				return errors.Wrapf(err, "copier: get: lstat %q", item)
+			}
+			// chase links. if we hit a dead end, we should just fail
+			followedLinks := 0
+			const maxFollowedLinks = 16
+			for info.Mode()&os.ModeType == os.ModeSymlink && followedLinks < maxFollowedLinks {
```

Comment:
Nothing around here prevents access outside of `root` AFAICS.

Why is this a separate implementation from `resolvePath`?

Shouldn’t this treat `Excludes` the same way that code does? The API documentation doesn’t suggest any difference between `Stat` and `Get` in that respect.